sdk/nodejs/runtime/closure/createClosure.ts

```diff
@@ -933,51 +933,58 @@ async function getOrCreateEntryAsync(
-            // we're about to recurse inside this object.  In order to prever infinite
-            // loops, put a dummy entry in the environment map.  That way, if we hit
-            // this object again while recursing we won't try to generate this property.
             const keyEntry = await getOrCreateEntryAsync(getNameOrSymbol(descriptor), undefined, context, serialize, logInfo);
-            if (!environment.has(keyEntry)) {
-                environment.set(keyEntry, <any>undefined);
 
-                const propertyInfo = await createPropertyInfoAsync(descriptor);
-                const prop = await getOwnPropertyAsync(obj, descriptor);
-                const valEntry = await getOrCreateEntryAsync(
-                    prop, undefined, context, serialize, logInfo);
-
-                // Now, replace the dummy entry with the actual one we want.
-                environment.set(keyEntry, { info: propertyInfo, entry: valEntry });
+            // We're about to recurse inside this object.  In order to prevent infinite loops, put a
+            // dummy entry in the environment map.  That way, if we hit this object again while
+            // recursing we won't try to generate this property.
+            //
+            // Note: we only stop recursing if we hit exactly our sentinel key (i.e. we're self
+            // recursive).  We *do* want to recurse through the object again if we see it through
+            // non-recursive paths.  That's because we might be hitting this object through one
+            // prop-name-path, but we created it the first time through another prop-name path.
+            //
+            // By processing the object again, we will add the different members we need.
+            if (object.env.has(keyEntry) && object.env.get(keyEntry) === undefined) {
+                continue;
             }
+            object.env.set(keyEntry, <any>undefined);
+
+            const propertyInfo = await createPropertyInfoAsync(descriptor);
+            const prop = await getOwnPropertyAsync(obj, descriptor);
+            const valEntry = await getOrCreateEntryAsync(
+                prop, undefined, context, serialize, logInfo);
+
+            // Now, replace the dummy entry with the actual one we want.
+            object.env.set(keyEntry, { info: propertyInfo, entry: valEntry });
         }
 
         // If the object's __proto__ is not Object.prototype, then we have to capture what it
         // actually is.  On the other end, we'll use Object.create(deserializedProto) to set
         // things up properly.
         //
         // We don't need to capture the prototype if the user is not capturing 'this' either.
-        if (!entry.object!.proto) {
+        if (!object.proto) {
             const proto = Object.getPrototypeOf(obj);
             if (proto !== Object.prototype) {
-                entry.object!.proto = await getOrCreateEntryAsync(
+                object.proto = await getOrCreateEntryAsync(
```

Comment:
i could avoid the null checks here because entry.proto is the value passed in, and is known to be non-null.